File: ruby-backend/app/models/track_vote.rb
# TrackVote model - Compatible with Laravel track_votes table
require 'sequel'
require 'securerandom'

class TrackVote < Sequel::Model
  # Use existing Laravel table structure
  set_dataset :track_votes
  
  # Plugins for enhanced functionality
  plugin :validation_helpers
  
  # Allow setting id for UUID compatibility (needed for tests)
  unrestrict_primary_key
  
  # UUID generation for primary key (Laravel compatibility)
  def before_create
    self.id ||= SecureRandom.uuid
    self.created_at ||= Time.now
    super
  end
  
  # Associations
  many_to_one :track, class: :Track, key: :track_id
  many_to_one :user, class: :User, key: :user_id
  
  # Validation rules (compatible with Laravel validation)
  def validate
    super
    validates_presence [:track_id, :user_id]
    validates_unique [:track_id, :user_id], message: 'user has already voted for this track'
  end
  
  # Class methods for validation rules (Laravel compatibility)
  def self.validation_rules
    {
      track_id: 'required|uuid|exists:tracks,id',
      user_id: 'required|uuid|exists:users,id'
    }
  end
  
  # Get room through track association (Laravel compatibility)
  def room
    track&.room
  end
  
  # Check if vote is from track uploader (Laravel compatibility)
  def from_uploader?
    user_id == track&.uploader_id
  end
  
  # Serialization for API responses (Laravel format compatibility)
  def to_hash
    {
      id: id,
      track_id: track_id,
      user_id: user_id,
      user: user&.to_hash,
      track: track&.to_hash,
      created_at: created_at&.iso8601,
      is_from_uploader: from_uploader?
    }
  end
  
  def to_json(*args)
    to_hash.to_json(*args)
  end
  
  # Hooks for model events (Laravel compatibility)
  def after_create
    # Update track vote score when vote is created
    track&.update(vote_score: track.vote_score + 1)
    super
  end
  
  def after_destroy
    # Update track vote score when vote is deleted
    track&.update(vote_score: [track.vote_score - 1, 0].max)
    super
  end
end

File: backend/app/Models/TrackVote.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Concerns\HasUuids;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Carbon\Carbon;

class TrackVote extends Model
{
    use HasFactory, HasUuids;

    protected $fillable = [
        'track_id',
        'user_id',
        'created_at',
    ];

    protected $casts = [
        'created_at' => 'datetime',
    ];

    public $timestamps = false;

    /**
     * Validation rules for track vote
     */
    public static function validationRules(): array
    {
        return [
            'track_id' => 'required|uuid|exists:tracks,id',
            'user_id' => 'required|uuid|exists:users,id',
        ];
    }

    /**
     * Track this vote belongs to
     */
    public function track(): BelongsTo
    {
        return $this->belongsTo(Track::class);
    }

    /**
     * User who cast this vote
     */
    public function user(): BelongsTo
    {
        return $this->belongsTo(User::class);
    }

    /**
     * Get the room this vote belongs to (through track)
     */
    public function room(): BelongsTo
    {
        return $this->track()->getRelated()->room();
    }

    /**
     * Check if this vote is from the track uploader
     */
    public function isFromUploader(): bool
    {
        return $this->user_id === $this->track->uploader_id;
    }

    /**
     * Boot method to handle model events
     */
    protected static function boot()
    {
        parent::boot();

        // Set created_at timestamp when creating
        static::creating(function ($vote) {
            if (!$vote->created_at) {
                $vote->created_at = Carbon::now();
            }
        });

        // Update track vote score when vote is created
        static::created(function ($vote) {
            $vote->track->increment('vote_score');
        });

        // Update track vote score when vote is deleted
        static::deleted(function ($vote) {
            $vote->track->decrement('vote_score');
        });
    }
}